enhancements/ingress/global-admission-hsts.md

```diff
@@ -0,0 +1,631 @@
+```
+To ease administration, cluster administrators can use this simple command-line script, run with an input file of routes:
+````shell
+#!/bin/bash
+
+# Usage: ./batch_annotate.sh filename.txt
+# Where filename.txt lines are formatted: route_name annotation_key annotaion_value namespace overwrite
+# For example: console  haproxy.router.openshift.io/hsts_header  max-age=600 openshift-console true
+
+CR=$'\n'
+# Annotate routes with a given route name (but will not overwrite existing)
+function annotate_route () {
+  local route_name=${1}
+  local annotation_key=${2}
+  local annotation_value=${3}
+  local namespace=${4}
+  local overwrite=${5}
+  echo "${CR}[cmd]: oc annotate route --overwrite=${overwrite} ${route_name} ${annotation_key}=${annotation_value} -n ${namespace}"
+  oc annotate route ${route_name} --overwrite=${overwrite} ${annotation_key}=${annotation_value} -n ${namespace}
+  return $?
+}
+
+function check_result () {
+  local r=${1}
+  if [[ $r -eq 0 ]]; then
+    echo "Success"
+  else
+    echo "Failure: error code $r"
+  fi
+}
+
+if [[ -z ${1} ]]; then
+  echo "Usage: must enter filename"
+  exit 1
+else
+  echo "Reading from ${1}..."
+fi
+
+while IFS= read -r line; do
+  IFS=' ' read -r -a fields <<< "$line"
+  annotate_route  ${fields[0]} ${fields[1]}  ${fields[2]}  ${fields[3]} ${fields[4]}
+  result=$?
+  check_result $result
+done < "${1}"
+
+echo "Done"
+````
+Input file example:
+```text
+console  haproxy.router.openshift.io/hsts_header  max-age=900 openshift-console true
+myBigRoute  haproxy.router.openshift.io/hsts_header  max-age=31536000;preload;includeSubdomains openshift-console true
+```
+Usage example:
+`./batch_annoation.sh filename.txt`
+
+Accompanying unit and e2e test changes will be added to exercise the `RequiredHSTSPolicies` type.  The API guarantees:
+* Routes with hosts that match the `domainPatterns` of the HSTS will be validated
+* Routes with hosts that match the `domainPatterns` and `namespaceSelector` of the HSTS will be validated
+* Routes that only match the `namespaceSelector` will not be validated
+* Routes that are validated will validate that the `maxAge` exists and falls within the range of the HSTS `maxAgePolicy`
+* Routes that are validated will validate that the `preloadPolicy` and `includeSubdomainPolicy` match the HSTS Policy if they exist
+
+#### Audit HSTS configurations on any cluster
+This proposal offers the cluster administrator a way to review the HSTS configurations guaranteed by the admission plugin.
+For example, to review the maxAge set for required HSTS Policies:
+```shell
+oc get clusteroperator/ingress -n openshift-ingress-operator -o jsonpath='{range .spec.requiredHSTSPolicies}{.spec.requiredHSTSPolicies.maxAgePolicy.largestMaxAge}{"\n"}{end}'
+```
+To review the annotations on all routes:
+````shell
+oc get route  --all-namespaces -o go-template='{{range .items}}Name: {{.metadata.name}} Annotations: {{.metadata.annotations}} {{"\n"}}{{end}}'
```

Comment:
Great example.  Can we get selective down to the `haproxy.router.openshift.io/hsts_header` annotation only?

Reply:
Updated this.

Reply:
I added this.